README.md
# JWT NaCl [![travis][ci_img]][travis] [![yard docs][yd_img]][yard_docs] [![code climate][cc_img]][code_climate]

## A JSON Web Token (JWT) implementation using NaCl cryptography

### Description
A Ruby JSON Web Token implementation using Edwards-curve Digital Signature Algorithm ([EdDSA][eddsa]) [curve Ed25519 digital signatures][ed25519] from the state-of-the-art NaCl [Networking and Cryptography library][nacl] by [Daniel J. Bernstein][bernstein].

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'jwt_nacl'
```

And then execute:

    $ bundle

Or install it directly as:

    $ gem install jwt_nacl

### Philosophy & Design Goals
* Convention over configuration
* Use of state-of-the-art cryptography, including EdDSA Curve25519 elliptic curves
* Minimal API surface area
* Thorough test coverage
* Modularity for comprehension and extensibility

### Why NaCl?
Cryptography typically exposes a high degree of complexity, due to many possible configuration decisions. One poor choice could result in an insecure system.

NaCl is different. NaCl provides an expertly-assembled, high-level cryptographic API with the correct configuration already built-in. See [RbNaCl][rbnacl] for more rationale.

For a more conventional JWT implementation, please refer to the related [json_web_token](https://github.com/garyf/json_web_token) gem.

## Usage

### JWT.sign(claims, private_key)

#### Returns a 3 element hash that includes:
* a JSON Web Token
* the private key
* the public key

`claims` (required) hash (non-empty)

`private_key` (optional) string, 32 random byte signing key

Example

```ruby
require 'jwt_nacl'

claims = {iss: "mike", exp: 1300819380, :"http://example.com/is_root" => false}

private_hex = "d2c5c54bc205266f12a8a21809aa2989536959f666a5d68710e6fab94674041a"
private_key = [private_hex].pack("H*")

public_hex = "1e10af4b79b8d005c8b4237161f1350844b2e6c1a8d6aa4817151c04a2751731"
public_key = [public_hex].pack("H*")

# Sign with an elliptical curve Ed25519 digital signature
jwt = JWT.sign(claims, private_key)
#=> {
  jwt: "eyJhbGciOiJFZDI1NTE5IiwidHlwIjoiSldUIn0.eyJpc3MiOiJtaWtlIiwiZXhwIjoxMzAwODE5MzgwLCJodHRwOi8vZXhhbXBsZS5jb20vaXNfcm9vdCI6ZmFsc2V9.f2y6Sax9eK9M3JiFCt4ZfzzOL56SWNhydHpPPIoVkm21D3_bJq5DmFLgH8ee2OlzSlZMoq009jLSg6AC0mn4DA",
  private_key: private_key,
  public_key: public_key
}
```

### JWT.verify(jwt, public_key)

#### Returns a hash:
* \{claims: < JWT claims set >\}, if the digital signature is verified
* \{error: "invalid"\}, otherwise

`jwt` (required) is a JSON web token string

`public_key` (required) string, 32 byte verifying key

Example

```ruby
require 'jwt_nacl'

jwt = "eyJhbGciOiJFZDI1NTE5IiwidHlwIjoiSldUIn0.eyJpc3MiOiJtaWtlIiwiZXhwIjoxMzAwODE5MzgwLCJodHRwOi8vZXhhbXBsZS5jb20vaXNfcm9vdCI6ZmFsc2V9.f2y6Sax9eK9M3JiFCt4ZfzzOL56SWNhydHpPPIoVkm21D3_bJq5DmFLgH8ee2OlzSlZMoq009jLSg6AC0mn4DA"

hex = "1e10af4b79b8d005c8b4237161f1350844b2e6c1a8d6aa4817151c04a2751731"
public_key = [hex].pack("H*")

# Verify with an elliptical curve Ed25519 public key
JWT.verify(jwt, public_key)
#=> {
  claims: {iss: "mike", exp: 1300819380, :"http://example.com/is_root" => false}
}
```

### Supported encryption algorithm
Ed25519, Edwards-curve Digital Signature Algorithm (EdDSA) using Curve25519

### Supported Ruby versions
Ruby 2.2.6 and up

[eddsa]: https://en.wikipedia.org/wiki/EdDSA
[ed25519]: http://ed25519.cr.yp.to/
[nacl]: http://nacl.cr.yp.to/
[bernstein]: https://en.wikipedia.org/wiki/Daniel_J._Bernstein

[rbnacl]: https://github.com/cryptosphere/rbnacl/blob/master/README.md
[travis]: https://travis-ci.org/garyf/jwt_nacl
[ci_img]: https://travis-ci.org/garyf/jwt_nacl.svg?branch=master
[yard_docs]: http://www.rubydoc.info/github/garyf/jwt_nacl
[yd_img]: http://img.shields.io/badge/yard-docs-blue.svg
[code_climate]: https://codeclimate.com/github/garyf/jwt_nacl
[cc_img]: https://codeclimate.com/github/garyf/jwt_nacl/badges/gpa.svg


CHANGELOG.md
## Changelog

### v0.1.0 (2017-01-29)

* Initial
  * support Edwards-curve Digital Signature Algorithm (EdDSA) using Curve25519


test/test_helper.rb
require "simplecov"
SimpleCov.start

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'jwt_nacl'

require 'minitest/autorun'
require "minitest/reporters"
Minitest::Reporters.use! [Minitest::Reporters::DefaultReporter.new]


test/jwt_nacl_test.rb
require "test_helper"
require "jwt_nacl/util"

describe JwtNacl do
  describe ".sign and .verify" do
    it "with private_key" do
      # seed = JwtNacl::Util.ed25519_random_seed
      # hex = seed.unpack("H*").first
      private_hex = "d2c5c54bc205266f12a8a21809aa2989536959f666a5d68710e6fab94674041a"
      private_key = [private_hex].pack("H*")
      claims = {iss: "mike", exp: 1300819380, :"http://example.com/is_root" => false}

      hsh_0 = JWT.sign(claims, private_key)
      jwt = hsh_0[:jwt]
      expect(jwt.split(".").length).must_equal 3
      expect(jwt).must_equal "eyJhbGciOiJFZDI1NTE5IiwidHlwIjoiSldUIn0.eyJpc3MiOiJtaWtlIiwiZXhwIjoxMzAwODE5MzgwLCJodHRwOi8vZXhhbXBsZS5jb20vaXNfcm9vdCI6ZmFsc2V9.f2y6Sax9eK9M3JiFCt4ZfzzOL56SWNhydHpPPIoVkm21D3_bJq5DmFLgH8ee2OlzSlZMoq009jLSg6AC0mn4DA"

      public_key = hsh_0[:public_key]
      public_hex = public_key.unpack("H*").first
      expect(public_hex).must_equal "1e10af4b79b8d005c8b4237161f1350844b2e6c1a8d6aa4817151c04a2751731"
      expect(public_key.length).must_equal 32

      hsh_1 = JWT.verify(jwt, public_key)
      expect(hsh_1[:claims]).must_equal claims
    end

    it "without private_key" do
      claims = {iss: "carol", exp: 1900723794, :"http://example.com/is_root" => true}
      hsh_0 = JWT.sign(claims)
      jwt = hsh_0[:jwt]
      expect(jwt.split(".").length).must_equal 3

      private_key = hsh_0[:private_key]
      expect(private_key.length).must_equal 32
      public_key = hsh_0[:public_key]
      expect(public_key.length).must_equal 32

      hsh_1 = JWT.verify(jwt, public_key)
      expect(hsh_1[:claims]).must_equal claims
    end
  end

  describe ".sign with invalid claims" do
    it "nil" do
      e = assert_raises(RuntimeError) { JWT.sign(nil) }
      assert_match(/invalid claims/, e.message)
    end

    it "an empty string" do
      e = assert_raises(RuntimeError) { JWT.sign("") }
      assert_match(/invalid claims/, e.message)
    end

    it "an empty hash" do
      e = assert_raises(RuntimeError) { JWT.sign(Hash.new) }
      assert_match(/invalid claims/, e.message)
    end

    it "a string" do
      e = assert_raises(RuntimeError) { JWT.sign("claims") }
      assert_match(/invalid claims/, e.message)
    end
  end
end


test/jwt_nacl/jws_test.rb
require "test_helper"
require "jwt_nacl/jws"
require "jwt_nacl/util"

module JwtNacl
  describe Jws do
    describe ".sign and .verify" do
      it "with private_key" do
        # seed = Util.ed25519_random_seed
        # hex = seed.unpack("H*").first
        hex = "52647b30343608659d043fc880ed8c39d5765d6e91117e58694ff2f8e66e940e"
        private_key = [hex].pack("H*")
        payload = '{"iss":"joe","exp":1300819380,"http://example.com/is_root":true}'

        hsh_0 = Jws.sign(payload, private_key)
        jwt = hsh_0[:jwt]
        expect(jwt.split(".").length).must_equal 3
        expect(hsh_0[:private_key]).must_equal private_key

        hsh_1 = Jws.verify(jwt, hsh_0[:public_key])
        expect(hsh_1[:jwt]).must_equal jwt
      end

      it "without private_key" do
        payload = '{"iss":"joe","exp":1300819380,"http://example.com/is_root":true}'
        hsh_0 = Jws.sign(payload)
        jwt = hsh_0[:jwt]
        expect(jwt.split(".").length).must_equal 3
        expect(hsh_0[:private_key].length).must_equal 32

        hsh_1 = Jws.verify(jwt, hsh_0[:public_key])
        expect(hsh_1[:jwt]).must_equal jwt
      end
    end
  end
end


lib/jwt_nacl/jws.rb
require "jwt_nacl/base64_url"
require "jwt_nacl/ed25519"

module JwtNacl
  # Represent content to be secured with digital signatures
  # @see http://tools.ietf.org/html/rfc7515
  module Jws
    HEADER = {
      alg: "Ed25519",
      typ: "JWT"
    }
    JWS_PARTS = 3

    module_function

    # @param payload [String] input for a digital signature computation
    # @param private_key [String] 32 random bytes (optional)
    # @return [Hash] a hash with a signed jwt, private_key, and public_key
    # @example
    #   Jws.sign(payload, private_key)
    #   # => {jwt: jwt, private_key: private_key, public_key: public_key}
    def sign(payload, private_key = nil)
      signing_input = encode_input(payload)
      hsh = Ed25519.sign(signing_input, private_key)
      {
        jwt: "#{signing_input}.#{Base64Url.encode(hsh[:signature])}",
        private_key: hsh[:private_key],
        public_key: hsh[:public_key]
      }
    end

    # @param jwt [String] input to be authenticated
    # @param public_key [String] 32 byte key used to authenticate a digital signature
    # @return [Hash] a hash with a verified jwt (e.g. jwt: jwt) or not verified (e.g. error: "invalid")
    # @example
    #   Jws.verify(jwt, public_key)
    #   # => {jwt: jwt}
    def verify(jwt, public_key)
      verified?(jwt, public_key) ? {jwt: jwt} : {error: "invalid"}
    end

    def encode_input(payload)
      "#{Base64Url.encode(HEADER.to_json)}.#{Base64Url.encode(payload)}"
    end

    def verified?(jwt, public_key)
      ary = jwt.split(".")
      return unless ary.length == JWS_PARTS
      signature = Base64Url.decode(ary[2])
      message = "#{ary[0]}.#{ary[1]}"

      Ed25519.verify(public_key, signature, message)
    end

    private_class_method :encode_input,
      :verified?
  end
end


lib/jwt_nacl/util.rb
require "rbnacl/libsodium"
require "rbnacl"

module JwtNacl
  # Utility methods
  module Util
    ED25519_SEEDBYTES = 32

    module_function

    def ed25519_random_seed
      RbNaCl::Random.random_bytes(ED25519_SEEDBYTES)
    end

    # @param hsh [Hash]
    # @return [Hash] a new hash with all keys converted to symbols,
    #   provided that they respond to .to_sym
    # @example
    #   Util.symbolize_keys({"a" =>  0, "b" => "2", c: "3"})
    #   # => {a: 0, b: "2", c: "3"}
    # @see cf. rails activesupport/lib/active_support/core_ext/hash/keys.rb
    def symbolize_keys(hsh)
      transform_keys(hsh) { |key| key.to_sym rescue key }
    end

    def transform_keys(hsh)
      result = Hash.new
      hsh.keys.each { |k| result[yield(k)] = hsh[k] }
      result
    end

    private_class_method :transform_keys
  end
end


lib/jwt_nacl/ed25519.rb
require "rbnacl/libsodium"
require "rbnacl"

module JwtNacl
  # Sign or verify JSON Web Signature (JWS) signing input using Edwards-curve
  # Digital Signature Algorithm (EdDSA) curve Ed25519
  # @see https://ed25519.cr.yp.to/
  module Ed25519
    module_function

    # @param message [String] input payload for a digital signature computation
    # @param seed [String] 32 random bytes (optional)
    # @return [Hash] a hash with the private_key, public_key, and signature
    # @example
    #   Ed25519.sign(message, seed)
    #   # => {private_key: private_key, public_key: public_key, signature: signature}
    def sign(message, seed = nil)
      signing_key = seed ? RbNaCl::SigningKey.new(seed) : RbNaCl::SigningKey.generate
      {
        private_key: signing_key.to_bytes,
        public_key: signing_key.verify_key.to_bytes,
        signature: signing_key.sign(message)
      }
    end

    # @param public_key [String] 32 byte key used to authenticate a digital signature
    # @param signature [String] alleged signature to be checked
    # @param message [String] message to be authenticated
    # @return [String, Boolean] verified message or false
    # @example
    #   Ed25519.verify(public_key, signature, message)
    #   # => message
    def verify(public_key, signature, message)
      verify_key(public_key, signature, message)
      message
    rescue
      false
    end

    def verify_key(public_key, signature, message)
      RbNaCl::VerifyKey.new(public_key)
        .verify(signature, message)
    end
  end
end


lib/jwt_nacl/base64_url.rb
require "base64"

module JwtNacl
  # Provide base64url encoding and decoding functions without padding, based upon standard
  # base64 encoding and decoding functions that do use padding
  # @see http://tools.ietf.org/html/rfc7515#appendix-C
  module Base64Url
    module_function

    # @param str [String]
    # @return [String] a urlsafe_encode64 string with all trailing "=" padding removed
    # @example
    #   Base64Url.encode("foo")
    #   # => "Zm9v"
    def encode(str)
      base64_padding_removed(Base64.urlsafe_encode64(str))
    end

    # @param str [String] encoded as url_encode64
    # @return [String] with trailing "=" padding added before decoding
    # @example
    #   Base64Url.decode("YmFy")
    #   # => "bar"
    def decode(str)
      Base64.urlsafe_decode64(base64_padding_added(str))
    end

    def base64_padding_removed(encoded)
      encoded.gsub(/[=]/, "")
    end

    def base64_padding_added(str)
      mod = str.length % 4
      return str if mod == 0
      raise("Invalid base64 string") if mod == 1
      "#{str}#{"=" * (4 - mod)}"
    end

    private_class_method :base64_padding_removed,
      :base64_padding_added
  end
end


lib/jwt_nacl.rb
require "jwt_nacl/jws"
require "jwt_nacl/util"

# Encode claims for transmission as a JSON object that is used as the payload
# of a JSON Web Signature (JWS) structure, enabling the claims
# to be integrity protected with a signature for later verification
# @see http://tools.ietf.org/html/rfc7519
module JwtNacl
  module_function

  # @param claims [Hash] input for a digital signature computation
  # @param private_key [String] 32 random bytes (optional)
  # @return [Hash] a hash with a signed jwt, private_key, and public_key
  # @example
  #   claims = {iss: "mike", exp: 1300819380, :"http://example.com/is_root" => false}
  #   private_hex = "d2c5c54bc205266f12a8a21809aa2989536959f666a5d68710e6fab94674041a"
  #   private_key = [private_hex].pack("H*")
  #   public_hex = "1e10af4b79b8d005c8b4237161f1350844b2e6c1a8d6aa4817151c04a2751731"
  #   public_key = [public_hex].pack("H*")
  #   JwtNacl.sign(claims, private_key)
  #   # => {jwt: "eyJhbGciOiJFZDI1NTE5IiwidHlwIjoiSldUIn0.eyJpc3MiOiJtaWtlIiwiZXhwIjoxMzAwODE5MzgwLCJodHRwOi8vZXhhbXBsZS5jb20vaXNfcm9vdCI6ZmFsc2V9.f2y6Sax9eK9M3JiFCt4ZfzzOL56SWNhydHpPPIoVkm21D3_bJq5DmFLgH8ee2OlzSlZMoq009jLSg6AC0mn4DA", private_key: private_key, public_key: public_key}
  def sign(claims, private_key = nil)
    Jws.sign(validated_payload(claims), private_key)
  end

  # @param jwt [String] a JSON Web Token
  # @param public_key [String] 32 byte verifying key
  # @return [Hash] +{claims: < the jwt claims set hash >}+ if the jwt verifies,
  #   or +{error: "invalid"}+ otherwise
  # @example
  #   jwt = "eyJhbGciOiJFZDI1NTE5IiwidHlwIjoiSldUIn0.eyJpc3MiOiJtaWtlIiwiZXhwIjoxMzAwODE5MzgwLCJodHRwOi8vZXhhbXBsZS5jb20vaXNfcm9vdCI6ZmFsc2V9.f2y6Sax9eK9M3JiFCt4ZfzzOL56SWNhydHpPPIoVkm21D3_bJq5DmFLgH8ee2OlzSlZMoq009jLSg6AC0mn4DA"
  #   hex = "1e10af4b79b8d005c8b4237161f1350844b2e6c1a8d6aa4817151c04a2751731"
  #   public_key = [hex].pack("H*")
  #   JwtNacl.verify(jwt, public_key)
  #   # => {claims: {iss: "mike", exp: 1300819380, :"http://example.com/is_root" => false}}
  def verify(jwt, public_key)
    verified_claims(Jws.verify(jwt, public_key))
  end

  def validated_payload(claims)
    raise("invalid claims") if !claims || claims.empty? || !claims.is_a?(Hash)
    claims.to_json
  end

  def verified_claims(hsh)
    return {error: "invalid"} if hsh[:error]
    {claims: decoded_claims(hsh[:jwt].split(".")[1])}
  end

  def decoded_claims(str)
    Util.symbolize_keys(
      JSON.parse(
        Base64Url.decode(str)
      )
    )
  end

  private_class_method :validated_payload,
    :verified_claims,
    :decoded_claims
end

# alias
JWT = JwtNacl
